streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala

```diff
@@ -123,6 +126,15 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
         timedOut
       }
 
+      // generate one more batch to make sure RDD in lastJob is checkpointed. As an performance
+      // optimization, if the latest info has been checkpointed in last batch, there is no need
+      // to run another round. "isCheckpointMissedLastTime" method here is in charge of collect
+      // such information from every DStream recursively.
+      if (!jobScheduler.receiverTracker.hasUnallocatedBlocks &&
+        ssc.graph.isCheckpointMissedLastTime) {
+        Thread.sleep(ssc.graph.batchDuration.milliseconds)
+      }
```

Comment:
I just wonder if we can add a double check if the last mini batch finish, otherwise, at least we can add a warning log.